ArjunaJTA/jta/classes/com/arjuna/ats/internal/jta/resources/arjunacore/XAResourceRecord.java

```diff
@@ -227,10 +225,15 @@ public int topLevelPrepare()
 			case XAException.XA_RBOTHER:
 			case XAException.XA_RBPROTO:
 			case XAException.XA_RBTIMEOUT:
+				// This will avoid calling rollback when aborting
+				_rolledBack = true;
+				return TwoPhaseOutcome.PREPARE_NOTOK;
+			// We might need to roll back the XA resource for these error codes
 			case XAException.XAER_INVAL:
 			case XAException.XAER_PROTO:
-			case XAException.XAER_NOTA: // resource may have arbitrarily rolled back (shouldn't, but ...)
-				return TwoPhaseOutcome.PREPARE_NOTOK;  // will not call rollback
+			case XAException.XAER_RMERR:
+			case XAException.XAER_RMFAIL:
+				return TwoPhaseOutcome.PREPARE_NOTOK;
```

Comment:
Maybe we need a comment something like:
// We do not know the outcome of the resource manager conclusively and so we will try to call rollback again

Reply:
Sorry, assuming that was there before I missed that. I will resolve this.